[SelectionDAG][RISCV] Avoid store merging across function calls

mikhailramalho opened the pull request:
This patch improves DAGCombiner's handling of potential store merges by
detecting function calls between loads and stores. When a function call
exists in the chain between a load and its corresponding store, we avoid
merging these stores as it would require costly register spilling.

We had to implement a hook on TLI, since TTI is unavailable in DAGCombine.

Currently, it's only enabled for riscv.

This is the DAG equivalent of PR #129258

mikhailramalho committed:
[SelectionDAG] Avoid store merging across function calls

This patch improves DAGCombiner's handling of potential store merges by
detecting function calls between loads and stores. When a function call
exists in the chain between a load and its corresponding store, we avoid
merging these stores as it would require costly register spilling.

Currently it's only enabled for riscv.

mikhailramalho committed:
Check call_end instead of start

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Only walk over mem operation chains

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Don't go over NumConsecutiveStores

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Use SDValues

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Added fallthrough

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Add Visited list to cache the walk

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Moved increment

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Updated test case

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Enable merge by default for scalars

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Rewrite walk back algo to keep track of calls found

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Check final type before we prevent merges

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

topperc commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -21553,6 +21553,56 @@ bool DAGCombiner::tryStoreMergeOfLoads(SmallVectorImpl<MemOpLink> &StoreNodes,
+              return FoundCall;
+            [[fallthrough]];
+          default:
+            if (Node->getNumOperands() > 0)
```
This getNumOperands check should only fail for EntryToken I think. Can we add EntryToken explicitly to the switch instead?

lukel97 replied:
Looks like we already have `Node->getNumOperands() == 0` in the continue at the top, is this check still needed?

topperc commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -21553,6 +21553,56 @@ bool DAGCombiner::tryStoreMergeOfLoads(SmallVectorImpl<MemOpLink> &StoreNodes,
+            [[fallthrough]];
+          default:
+            if (Node->getNumOperands() > 0)
+              Worklist.emplace_back(Node->getOperand(0).getNode(), FoundCall);
```
Assert that `Node->getOperand(0)` has MVT::Other type?

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -21553,6 +21553,56 @@ bool DAGCombiner::tryStoreMergeOfLoads(SmallVectorImpl<MemOpLink> &StoreNodes,
       JointMemOpVT = EVT::getIntegerVT(Context, SizeInBits);
     }
 
+    auto HasCallInLdStChain = [](SmallVectorImpl<MemOpLink> &StoreNodes,
```
Can we pull this out into a static function to make it similar to the `checkMergeStoreCandidatesForDependencies` check?

mikhailramalho replied:
Sure, do you want a static function or a member function like `checkMergeStoreCandidatesForDependencies`? 

Just asking because  `MemOpLink` is a private struct inside the `DAGCombiner` class, so we would need to make it public to pass it as argument of the static function.

lukel97 replied:
Oh I didn't realise it was a member function, but yeah a member function sounds fine

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -21553,6 +21553,56 @@ bool DAGCombiner::tryStoreMergeOfLoads(SmallVectorImpl<MemOpLink> &StoreNodes,
+
+        SmallPtrSet<const SDNode *, 32> Visited;
+        SmallVector<std::pair<const SDNode *, bool>, 8> Worklist;
+        Worklist.emplace_back(St->getOperand(0).getNode(), false);
```
Nit
```suggestion
        Worklist.emplace_back(St->getChain().getNode(), false);
```

lukel97 replied:
Isn't St a StoreSDNode?

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -21553,6 +21553,56 @@ bool DAGCombiner::tryStoreMergeOfLoads(SmallVectorImpl<MemOpLink> &StoreNodes,
+                                 SmallVectorImpl<MemOpLink> &LoadNodes,
+                                 unsigned NumStores) {
+      for (unsigned i = 0; i < NumStores; ++i) {
+        StoreSDNode *St = cast<StoreSDNode>(StoreNodes[i].MemNode);
+        SDValue Val = peekThroughBitcasts(St->getValue());
+        LoadSDNode *Ld = cast<LoadSDNode>(Val);
+        assert(Ld == LoadNodes[i].MemNode && "Load and store mismatch");
```
Nit, can we remove the peekThroughBitcasts in release builds if we swap the assertion around? Something like

```suggestion
        StoreSDNode *St = cast<StoreSDNode>(StoreNodes[i].MemNode);
        StoreSDNode *Ld = cast<LoadSDNode>(LoadNodes[i].MemNode);
        assert(Ld == peekThroughBitcasts(St->getValue()) && "Load and store mismatch");
```

mikhailramalho replied:
how about the following?
```
-        SDValue Val = peekThroughBitcasts(St->getValue());
-        LoadSDNode *Ld = cast<LoadSDNode>(Val);
-        assert(Ld == LoadNodes[i].MemNode && "Load and store mismatch");
+        LoadSDNode *Ld = cast<LoadSDNode>(LoadNodes[i].MemNode);
+        assert(Ld == cast<LoadSDNode>(peekThroughBitcasts(St->getValue())) &&
+               "Load and store mismatch");
```

lukel97 replied:
Argh woops, yes Ld should be LoadSDNode not StoreSDNode!

lukel97 commented on llvm/test/CodeGen/RISCV/stores-of-loads-merging.ll:
Can we add a negative test so to make sure that we still merge consecutive stores into scalars? E.g. two i32s into an i64

mikhailramalho committed:
No need to check operands. It's checked in the start of the loop

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Assert operand type

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Moved peekThroughBitcasts into an assertion

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

arsenm commented:
> This patch improves DAGCombiner's handling of potential store merges by
detecting function calls between loads and stores. 

I'm surprised this even happens. What happens if you just do this unconditionally?

mikhailramalho committed:
Use getChain instead of accessing the operand 0

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho commented:
> > This patch improves DAGCombiner's handling of potential store merges by
> > detecting function calls between loads and stores.
> 
> I'm surprised this even happens. What happens if you just do this unconditionally?

We want to still allow merging scalars... If we prevent merges unconditionally, we would miss that

Luke requested a test case for when we merge scalars, I'm working on it now and will add it to the PR soon

mikhailramalho committed:
Make hasCallInLdStChain a member function

mikhailramalho commented:
> I chatted a bit with @mikhailramalho offline, just recording the key bits.
> 
> 1. As mentioned by Luke above, we probably want to avoid inhibiting scalar merging.  Similarly, if we already have two small vectors combining into one larger vector is probably at least neutral.  (These are RISCV specific profitability statements.)
> 2. I would expect to see something in the check in terms of the callee saved registers for the callee of the call.  The _reason_ that scalar to vector merging is likely unprofitable is that we have scalar CSRs in the standard calling convention, but do not have vector CSRs.
> 3. A minimal patch might simply check if the proposed merged type has any CSRs.  That would cover the most painful case today.  A more complicate heuristic would try to reason about whether the merged type was a different register class than the source types, and then apply the CSR check.  We could get fancy when both register classes have CSRs, but the destination has fewer, but I suggest we defer that case until someone has an example which cares.
> 4. We do appear to have calls to getRegClassFor in DAGCombine already, so that knowledge should be available.  I don't see us currently using information about the calling convention in this code, hopefully that's available?  I haven't checked.
> 
> Edit: Re-reading this, I think I may have accidental mislead here. We could have the callee saved heuristic in DAG as mentioned, but we could also just extend the TLI hook you added slightly to have the source type, and have this reasoning explained in a comment in the hook. The result is static for the target.

I mentioned earlier to have this as a follow-up patch but maybe it's a small change and we can fit in this PR? I'll take a look at how to extend the TLI hook to check for callee save regs

mikhailramalho committed:
Added test case

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

preames commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -21553,6 +21601,15 @@ bool DAGCombiner::tryStoreMergeOfLoads(SmallVectorImpl<MemOpLink> &StoreNodes,
       JointMemOpVT = EVT::getIntegerVT(Context, SizeInBits);
     }
 
+    // Check if there is a call in the load/store chain.
```
I just noticed there is a better integration here.  The way you have this phrased, we either can merge to the vector type, or we don't merge at all.

I think what you actually want to do here is to add a couple parameters to the existing canMergeStoresTo interface.  If you add both the SrcVT, and a "IsOverCall" boolean, then that interface returning the answer suggested above in the IsOverCall case will result in a wider scalar type being chosen if one is available.

preames replied:
Though, I'll note I'm also fine in this being a follow up change.  I'm happy to pick that up if you want to focus on just getting this in.

mikhailramalho replied:
I'm happy with either option: to land this PR and do a follow-up patch, or work on this one. Up to you.

Also, I did a couple of tests with your first suggestion on checking for CSR for a given type in https://github.com/mikhailramalho/llvm-project/commit/df27cb2a90aeb62cce46ddac28fbee88b698926f, should I give up on that approach? There are still a couple of crashes I need to debug.

preames replied:
Let's go for minimum viable patch.  Update the existing TLI hook you added with the source parameter, let's get that landed, and then iterate.

mikhailramalho committed:
Merge remote-tracking branch 'origin/main' into dag-spillcost-fix

mikhailramalho committed:
Removed duplicated test after merge

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
No need to declare intrinsics anymore

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Removed unused args

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Address comment

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -21152,6 +21158,48 @@ bool DAGCombiner::checkMergeStoreCandidatesForDependencies(
+        break;
+      }
+    }
+    return false;
```
Should this return false be removed? Otherwise we never check anything but the first load/store pair

preames replied:
Hm, thinking about it, not sure we need to check any of the others.   (As in, we don't need the loop, and can only check the first pair.)  My reasoning is that for a call to be before LD2 and ST2, but not LD1 and ST2, then the call has to be between either LD1 and LD2 or ST1 and ST2.  In either case, either the call aliases the memory operation (preventing the merging entirely), or we can insert the merged LD or ST on whichever side of the call we prefer avoiding the issue.  

I haven't glanced at the code to see if it actually takes advantage of that rescheduling.  A couple of additional test cases coming...

preames commented on llvm/test/CodeGen/RISCV/stores-of-loads-merging.ll:
```diff
@@ -1,7 +1,6 @@
-declare i32 @llvm.experimental.constrained.fptosi.i32.f64(double, metadata)
 declare void @g()
 
 ; TODO: Merging scalars into vectors is unprofitable because we have no
```
Remember to update the TODO

mikhailramalho committed:
Address comment

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Removed todo

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

preames reviewed (approved):
LGTM

lukel97 reviewed (approved):
LGTM to me too. It would be nice to update hasCallInLdStChain to only check the first load/store pair but that's an optimisation, feel free to leave it to another PR if you want

mikhailramalho committed:
Simplify interface

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Merge remote-tracking branch 'origin/main' into dag-spillcost-fix

mikhailramalho commented:
> LGTM to me too. It would be nice to update hasCallInLdStChain to only check the first load/store pair but that's an optimisation, feel free to leave it to another PR if you want

It's a small change, so I'll update the PR and re-run the tests just to be sure

mikhailramalho committed:
Remove assert that fails when building blender_r

This asserts fails because the order of the store/loads are reversed
prior to the check, if we can rotate the loads.

topperc commented:
Why does the assert fail?

mikhailramalho commented:
> Why does the assert fail?

This line here swaps nodes: https://github.com/llvm/llvm-project/blob/d6c848db4a97103e19c398848bde0876a2c4c88e/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L21487

It happens when building `526.blender_r/src/blender/source/blender/blenloader/intern/readfile.c`

topperc commented:
> > Why does the assert fail?
> 
> This line here swaps nodes: https://github.com/llvm/llvm-project/blob/d6c848db4a97103e19c398848bde0876a2c4c88e/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L21487
> 
> It happens when building `526.blender_r/src/blender/source/blender/blenloader/intern/readfile.c`

If I remember correctly, the NeedRotate flag prevents the code from using vectors doesn't it?

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.h:
```diff
@@ -1070,6 +1070,13 @@ class RISCVTargetLowering : public TargetLowering {
+  /// calls between the load and store, since these are more expensive than just
+  /// using scalars
+  bool shouldMergeStoreOfLoadsOverCall(EVT SrcVT, EVT MergedVT) const override {
+    return SrcVT.isScalarInteger() == MergedVT.isScalarInteger();
```
I think SrcVT can be a scalar FP and MergeVT can be a scalar integer VT. In that case it would still be ok to merge across the call.

Maybe this should be

```
  return !MergedVT.isVector() || SrcVT.isVector();
```

preames replied:
I glanced at what we do for this today.  Quick summary:
2 x half (no zvfh) -> no merge performed
2 x half (zvfh) -> <2 x half> vector result type
2 x float -> <2 x float> vector result type
2 x float + noimplicitfloat -->  two integer loads, no merge

So basically, I think the patch as it is should be a net improvement (by disabling case 2 and 3).  I'm not opposed to your suggest change, just pointing out it's not needed to avoid a regression.

topperc replied:
> I glanced at what we do for this today. Quick summary:
> 2 x half (no zvfh) -> no merge performed
> 2 x half (zvfh) -> <2 x half> vector result type
> 2 x float -> <2 x float> vector result type
> 2 x float + noimplicitfloat --> two integer loads, no merge

@mikhailramalho said the case he hit the assert on was 2 x float -> i64.

> Note that the SDAG code doesn't even try to find a wider legal floating point type to merge to. The merge type will only be integer or vector.

A wider legal FP doesn't really make sense. You can't really join them without going through integer.

I think either `!MergedVT.isVector() || SrcVT.isVector()` or `MergedVT.isVector() == SrcVT.isVector()` would avoid cover vector without picking scalar FP in scalar integer.

preames replied:
Test cases (aside from the noimplicitfloat one) added in a5c7f81a

mikhailramalho commented:
> > > Why does the assert fail?
> > 
> > 
> > This line here swaps nodes: https://github.com/llvm/llvm-project/blob/d6c848db4a97103e19c398848bde0876a2c4c88e/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L21487
> > It happens when building `526.blender_r/src/blender/source/blender/blenloader/intern/readfile.c`
> 
> If I remember correctly, the NeedRotate flag prevents the code from using vectors doesn't it? Using a ROTL would only work for scalars.

Indeed, the code that triggers the assert fail is trying to merge two f32 into an i64

mikhailramalho committed:
Address comment

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Merge remote-tracking branch 'origin/main' into dag-spillcost-fix

mikhailramalho committed:
Update test

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Removed todo

Signed-off-by: Mikhail R. Gadelha <mikhail@igalia.com>

mikhailramalho committed:
Merge remote-tracking branch 'origin/main' into dag-spillcost-fix

preames reviewed (approved):
LGTM

topperc reviewed (approved):
LGTM
